# CS 370 Module Two Assignment
---
<div class="alert alert-block alert-success" style="color:black;">
<b>To Begin:</b> Use this <b>LastName_FirstName_Assignment2.ipynb</b> file to complete your assignment. Look for the section header <b>Start of Rubric Graded Items</b> for the placeholders for your assignment.
<br><br>
<b>When Finished:</b> When you have completed your work, download the project as an HTML file and submit the HTML file. Make sure that your file is named as specified in the rubric, and that you have addressed all of the rubric criteria in your response.
</div>

<div class="alert alert-block alert-info" style="color:black;">
    <b>Hint:</b> The code samples from the <b>Deep Learning with Keras</b> required readings pages 16-21 have been pre-loaded for you in the <b>Example_Code_Assignment2.ipynb</b> file. Some of the foundational starter code has also been added here as a starting point for your assignment as applicable.
</div>

<div class="alert alert-block alert-warning" style="color:black;">
    <b>Note:</b> For compatability purposes, libraries have been updated from those used in the required readings to match to current versions; hence some of the package invocations may differ slightly from the book. The affected lines of code have comments added to the right as applicable, with the old code commented out above for reference.
</div>

<div class="alert alert-block alert-danger" style="color:black;">
<b>GPU/CUDA/Memory Warnings/Errors:</b> You may receive some errors referencing that GPUs will not be used, CUDA could not be found, or free system memory allocation errors. These and a few others, are standard errors that can be ignored here as they are environment based.<br><br>
<b>Example messages:</b>
    <ul>
        <li>Could not find cuda drivers on your machine, GPU will not be used.</li>
        <li>Please check linkage and avoid linking the same target more than once.</li>
        <li>E external/local_xla/xla/stream_executor/cuda/cuda_platform.cc:51] failed call to cuInit: INTERNAL: CUDA error: Failed call to cuInit: UNKNOWN ERROR (303)</li>
        <li>Allocation of ######## exceeds 10% of free system memory</li>
    </ul>
</div>

---

## Required Baseline Starter Code
---

### Installing Required Packages
This is to install necessary components to run the assignment
<div class="alert alert-block alert-info" style="color:black;">
    <b>Hint:</b> Ensure that you run all starter code blocks before running your code from the <b>Start of Rubric Graded Items</b> sections.
</div>

In [1]:
!pip install -r requirements.txt


[notice] A new release of pip is available: 24.1.2 -> 26.0.1
[notice] To update, run: pip install --upgrade pip


---
### Loading the Required Libraries
<div class="alert alert-block alert-danger" style="color:black;">
<b>Warnings:</b> As a reminder, you may receive some errors here, these are standard errors that can be ignored as they are environment based.
</div>

Potentially helpful links: 
- https://keras.io/
- https://keras.io/api/datasets/
- https://keras.io/api/models/
- https://keras.io/api/layers/
- https://keras.io/api/optimizers/
- https://keras.io/api/utils/python_utils/

In [2]:
from __future__ import print_function
import numpy as np
from keras.datasets import mnist
from keras.models import Sequential
#from keras.layers.code import Dense, Activation # From the book, code replaced below
from keras.layers import Dense, Activation, Input # syntax change due to version bump
from keras.optimizers import SGD
#from keras.utils import np_utils # From the book, code replaced below
from keras.utils import to_categorical # syntax change due to version bump

---
### Generating a Numpy Random Seed For Reproducibility
<div class="alert alert-block alert-warning" style="color:black;">
    <b>Note:</b> The best practice is to not reseed a BitGenerator, but rather to recreate a new one. While this method for the the seed generation is considered a legacy function, here we use it to as a way to ensure reproducible results for repeatble course submissions.
</div>

Potentially helpful links: 
- https://keras.io/api/utils/python_utils/

In [3]:
np.random.seed(1671) # for reproducibility

---
### Network and Training Parameters
<div class="alert alert-block alert-danger" style="color:black;">
<b>Warnings:</b> As a reminder, you may receive some errors here, these are standard errors that can be ignored as they are environment based.
</div>

In [4]:
NB_EPOCH = 20
BATCH_SIZE = 128
VERBOSE = 1
NB_CLASSES = 10 # number of outputs = number of digits
OPTIMIZER = SGD() # SGD optimizer, explained later in this chapter
N_HIDDEN = 128
VALIDATION_SPLIT = 0.2 # how much TRAIN is reserved for VALIDATION

I0000 00:00:1773114531.893878   13860 cuda_executor.cc:1015] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
I0000 00:00:1773114531.949044   13860 cuda_executor.cc:1015] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
I0000 00:00:1773114531.952488   13860 cuda_executor.cc:1015] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
I0000 00:00:1773114531.956884   13860 cuda_executor.cc:1015] successful NUMA node read from SysFS ha

---
### Spliting Data Into Train and Test Sets

In [5]:
# data: shuffled and split between train and test sets
(X_train, y_train), (X_test, y_test) = mnist.load_data()

---
### Shaping

In [6]:
# X_train is 60000 rows of 28x28 values --> reshaped in 60000 x 784

RESHAPED = 784

X_train = X_train.reshape(60000, RESHAPED)
X_test = X_test.reshape(10000, RESHAPED)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')

In [7]:
# normalize
X_train /= 255
X_test /= 255

print(X_train.shape[0], 'train samples')
print(X_test.shape[0], 'test samples')

60000 train samples
10000 test samples


In [8]:
# convert class vectors to binary class matrices

# Y_train = np_utils.to_categorical(y_train, NB_CLASSES) # From the book, code replaced below
Y_train = to_categorical(y_train, NB_CLASSES) # Updated for current version

#Y_test = np_utils.to_categorical(y_test, NB_CLASSES) # From the book, code replaced below
Y_test = to_categorical(y_test, NB_CLASSES) # Updated for current version

---
# Start of Rubric Graded Items
---
<div class="alert alert-block alert-success" style="color:black;">
<b>To Begin:</b> Once you have reviewed the output of the blocks above, use the blocks below to complete your assignment portions.
</div>

<div class="alert alert-block alert-info" style="color:black;">
    <b>Hint:</b> As a reminder, the code samples from the <b>Deep Learning with Keras</b> required readings pages 16-21 have been pre-loaded for you in the <b>Example_Code_Assignment2.ipynb</b> file. Some of the foundational starter code has also been added here as a starting point for your assignment as applicable in the cells above.
</div>

---

## Notebook Configuration and Naming Convention
Click on LastName_FirstName_Assignment2.ipynb in the filetree and rename it following that convention. For example, if your name is Jane Doe, you would rename the file Doe_Jane_Assignment2.ipynb for later submission.

---
## Establishes Accuracy Scores with Two Hidden Layers
Follow the examples in the book to establish an accuracy rate for the training, validation, and test data sets with two hidden layers.


<div class="alert alert-block alert-warning" style="color:black;">
    <b>Note:</b> The example model configurations provided below must be modified for your assignment submission. Some lines are standard commands.
</div>

<div class="alert alert-block alert-info" style="color:black;">
<b>Hint:</b> Read through the MNIST handwritten digits example on pages 16-24 of <b>Deep Learning with Keras</b>. The code from pages 16-18 have been copied over to this Jupyter Notebook file for you as applicable in the cells above above for the minimum required baseline code. The hidden layers examples are provided starting on page 22, ensure to keep example modifications from above in mind when building out the rest of your assignment code blocks. The code below can be copied and used to start building out your hidden layers task.
<br><br>
<div style="background:#eff0f1;padding-left:10px">
<code style="margin-left:-5px"><b>Book Example Code Snippet: A simple neural net in Keras</b>
# 10 outputs
# final stage is softmax
model = Sequential()
# model.add(Dense(NB_CLASSES, input_shape=(RESHAPED,))) # From the book, code replaced below
model.add(Input(shape=(RESHAPED,))) # Updated for current version
model.add(Dense(NB_CLASSES)) # Updated for current version
model.add(Activation('softmax'))
model.summary()
model.compile(loss='categorical_crossentropy', optimizer=OPTIMIZER, metrics=['accuracy'])
history = model.fit(X_train, Y_train, batch_size=BATCH_SIZE, epochs=NB_EPOCH, verbose=VERBOSE,
validation_split=VALIDATION_SPLIT)
score = model.evaluate(X_test, Y_test, verbose=VERBOSE)
print("Test Score:", score[0])
print("Test Accuracy:", score[1])
</code>
</div>
</div>

<div class="alert alert-block alert-danger" style="color:black;">
<b>Warnings:</b> As a reminder, you may receive some errors in the model here, these are standard errors that can be ignored as they are environment based.
</div>

Potentially helpful links: 
- https://keras.io/api/models/sequential/
- https://keras.io/api/layers/activations/
- https://keras.io/api/models/sequential/#add-method
- https://keras.io/api/layers/
- https://keras.io/api/layers/core_layers/
- https://keras.io/api/models/model_training_apis/#compile-method
- https://keras.io/api/models/model_training_apis/#fit-method
- https://keras.io/api/models/model_training_apis/#evaluate-method

In [9]:
NB_EPOCH = 20
BATCH_SIZE = 128
OPTIMIZER = SGD()

model = Sequential()
model.add(Input(shape=(RESHAPED,)))
model.add(Dense(N_HIDDEN, activation='relu'))
model.add(Dense(N_HIDDEN, activation='relu'))
model.add(Dense(NB_CLASSES, activation='softmax'))

model.compile(
    loss='categorical_crossentropy',
    optimizer=OPTIMIZER,
    metrics=['accuracy'],
    run_eagerly=True
)

history = model.fit(
    X_train,
    Y_train,
    batch_size=BATCH_SIZE,
    epochs=NB_EPOCH,
    verbose=VERBOSE,
    validation_split=VALIDATION_SPLIT
)

score = model.evaluate(X_test, Y_test, verbose=VERBOSE)
print("Test accuracy:", score[1])

Epoch 1/20
375/375 ━━━━━━━━━━━━━━━━━━━━ 12s 29ms/step - accuracy: 0.4806 - loss: 1.8297 - val_accuracy: 0.8482 - val_loss: 0.6967
Epoch 2/20
375/375 ━━━━━━━━━━━━━━━━━━━━ 11s 28ms/step - accuracy: 0.8488 - loss: 0.6300 - val_accuracy: 0.8847 - val_loss: 0.4436
Epoch 3/20
375/375 ━━━━━━━━━━━━━━━━━━━━ 11s 28ms/step - accuracy: 0.8806 - loss: 0.4459 - val_accuracy: 0.9002 - val_loss: 0.3714
Epoch 4/20
375/375 ━━━━━━━━━━━━━━━━━━━━ 11s 29ms/step - accuracy: 0.8915 - loss: 0.3828 - val_accuracy: 0.9065 - val_loss: 0.3374
Epoch 5/20
375/375 ━━━━━━━━━━━━━━━━━━━━ 11s 30ms/step - accuracy: 0.9037 - loss: 0.3434 - val_accuracy: 0.9107 - val_loss: 0.3140
Epoch 6/20
375/375 ━━━━━━━━━━━━━━━━━━━━ 11s 29ms/step - accuracy: 0.9078 - loss: 0.3267 - val_accuracy: 0.9151 - val_loss: 0.2969
Epoch 7/20
375/375 ━━━━━━━━━━━━━━━━━━━━ 11s 29ms/step - accuracy: 0.9113 - loss: 0.3079 - val_accuracy: 0.9202 - val_loss: 0.2839
Epoch 8/20
375/375 ━━━━━━━━━━━━━━━━━━━━ 11s 30ms/step - accuracy: 0.9135 - loss: 0.2975 - 

---
## Establishes Accuracy Scores by Varying Parameters
Pick one parameter and run two or three different experiments modifying the parameter values to establish accuracy scores with different parameter values. Make sure the experiments result in significant changes in accuracy rates. Be sure to place each experiment in a different code block so that your instructor can view all of your changes.

<div class="alert alert-block alert-warning" style="color:black;">
    <b>Note:</b> You may have to do some research beyond the information provided in the book to implement these changes. Remember to copy your model definition, compilation, training, and evaluation blocks to test your experiment.
</div>

<div class="alert alert-block alert-info" style="color:black;">
<b>Hint:</b> Sample parameters that you can overwrite here were used in the <b>Network and Training Parameters</b> section above. This is not all inclusive and just provided as a baseline. See the additional example parameters starting on page 25 in the <b>Deep Learning with Keras</b> book.
<br><br><b>Important:</b> When codifying your experiements, ensure that you include the replacement variable in the cell below, followed by the hidden layers code you developed in the rubric item above.<br><br>
<div style="background:#eff0f1;padding-left:10px">
<code style="margin-left:-5px"><b>Book Example Code Snippet: Copied from cell above for quick reference to variable naming</b>
NB_EPOCH = 20
BATCH_SIZE = 128
VERBOSE = 1
NB_CLASSES = 10 # number of outputs = number of digits
OPTIMIZER = SGD() # SGD optimizer, explained later in this chapter
N_HIDDEN = 128
VALIDATION_SPLIT = 0.2 # how much TRAIN is reserved for VALIDATION
</code>
</div>
</div>

Potentially helpful links: 
- https://keras.io/api/models/model_training_apis/#fit-method
 
### Experiment 1

<div class="alert alert-block alert-info" style="color:black;">
<b>Hint:</b> When modifying parameters that will impact the optimization of the solution, ensure that you add the <b>OPTIMIZER = SGD()</b> line before your code with the hidden layers from above.
</div>

In [10]:
NB_EPOCH = 5
BATCH_SIZE = 128
OPTIMIZER = SGD()

model = Sequential()
model.add(Input(shape=(RESHAPED,)))
model.add(Dense(N_HIDDEN, activation='relu'))
model.add(Dense(N_HIDDEN, activation='relu'))
model.add(Dense(NB_CLASSES, activation='softmax'))

model.compile(
    loss='categorical_crossentropy',
    optimizer=OPTIMIZER,
    metrics=['accuracy'],
    run_eagerly=True
)

history = model.fit(
    X_train,
    Y_train,
    batch_size=BATCH_SIZE,
    epochs=NB_EPOCH,
    verbose=VERBOSE,
    validation_split=VALIDATION_SPLIT
)

score = model.evaluate(X_test, Y_test, verbose=VERBOSE)
print("Test accuracy:", score[1])

Epoch 1/5
375/375 ━━━━━━━━━━━━━━━━━━━━ 12s 32ms/step - accuracy: 0.4436 - loss: 1.8716 - val_accuracy: 0.8468 - val_loss: 0.7381
Epoch 2/5
375/375 ━━━━━━━━━━━━━━━━━━━━ 12s 32ms/step - accuracy: 0.8472 - loss: 0.6617 - val_accuracy: 0.8857 - val_loss: 0.4455
Epoch 3/5
375/375 ━━━━━━━━━━━━━━━━━━━━ 12s 32ms/step - accuracy: 0.8799 - loss: 0.4560 - val_accuracy: 0.8975 - val_loss: 0.3694
Epoch 4/5
375/375 ━━━━━━━━━━━━━━━━━━━━ 12s 32ms/step - accuracy: 0.8935 - loss: 0.3832 - val_accuracy: 0.9063 - val_loss: 0.3315
Epoch 5/5
375/375 ━━━━━━━━━━━━━━━━━━━━ 12s 32ms/step - accuracy: 0.9031 - loss: 0.3446 - val_accuracy: 0.9132 - val_loss: 0.3082
313/313 ━━━━━━━━━━━━━━━━━━━━ 6s 18ms/step - accuracy: 0.8990 - loss: 0.3538
Test accuracy: 0.9117000102996826


### Experiment 2
<div class="alert alert-block alert-info" style="color:black;">
<b>Hint:</b> When modifying parameters that will impact the optimization of the solution, ensure that you add the <b>OPTIMIXER = SGD()</b> line before your code with the hidden layers from above.
</div>

In [11]:
NB_EPOCH = 40
BATCH_SIZE = 128
OPTIMIZER = SGD()

model = Sequential()
model.add(Input(shape=(RESHAPED,)))
model.add(Dense(N_HIDDEN, activation='relu'))
model.add(Dense(N_HIDDEN, activation='relu'))
model.add(Dense(NB_CLASSES, activation='softmax'))

model.compile(
    loss='categorical_crossentropy',
    optimizer=OPTIMIZER,
    metrics=['accuracy'],
    run_eagerly=True
)

history = model.fit(
    X_train,
    Y_train,
    batch_size=BATCH_SIZE,
    epochs=NB_EPOCH,
    verbose=VERBOSE,
    validation_split=VALIDATION_SPLIT
)

score = model.evaluate(X_test, Y_test, verbose=VERBOSE)
print("Test accuracy:", score[1])

Epoch 1/40
375/375 ━━━━━━━━━━━━━━━━━━━━ 12s 32ms/step - accuracy: 0.4421 - loss: 1.8861 - val_accuracy: 0.8364 - val_loss: 0.7477
Epoch 2/40
375/375 ━━━━━━━━━━━━━━━━━━━━ 12s 31ms/step - accuracy: 0.8403 - loss: 0.6752 - val_accuracy: 0.8852 - val_loss: 0.4513
Epoch 3/40
375/375 ━━━━━━━━━━━━━━━━━━━━ 12s 31ms/step - accuracy: 0.8766 - loss: 0.4661 - val_accuracy: 0.8974 - val_loss: 0.3716
Epoch 4/40
375/375 ━━━━━━━━━━━━━━━━━━━━ 12s 31ms/step - accuracy: 0.8932 - loss: 0.3849 - val_accuracy: 0.9069 - val_loss: 0.3344
Epoch 5/40
375/375 ━━━━━━━━━━━━━━━━━━━━ 12s 33ms/step - accuracy: 0.9029 - loss: 0.3521 - val_accuracy: 0.9122 - val_loss: 0.3106
Epoch 6/40
375/375 ━━━━━━━━━━━━━━━━━━━━ 12s 32ms/step - accuracy: 0.9113 - loss: 0.3183 - val_accuracy: 0.9183 - val_loss: 0.2913
Epoch 7/40
375/375 ━━━━━━━━━━━━━━━━━━━━ 12s 32ms/step - accuracy: 0.9123 - loss: 0.3095 - val_accuracy: 0.9226 - val_loss: 0.2776
Epoch 8/40
375/375 ━━━━━━━━━━━━━━━━━━━━ 12s 32ms/step - accuracy: 0.9184 - loss: 0.2884 - 

### Experiment 3 (Optional)
<div class="alert alert-block alert-info" style="color:black;">
<b>Hint:</b> When modifying parameters that will impact the optimization of the solution, ensure that you add the <b>OPTIMIZER = SGD()</b> line before your code with the hidden layers from above.
</div>

---
## Explains Changes in Accuracy Rates
Create a Markdown cell in your Jupyter Notebook after your code and its outputs.
In this cell, explain the changes in accuracy rates by comparing and contrasting your results from Steps 3 and 4. What happens to the accuracy rates for the training, validation, and test data sets as you change the parameters. Why?

<div class="alert alert-block alert-info" style="color:black;">
    <b>Hint:</b> To use the markdown block below, double click in the <b>Type Markdown and LaTeX:  𝛼2</b> block below, to turn it back to html, Run the cell.
</div>

Potentially helpful links: 
- https://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Working%20With%20Markdown%20Cells.html

## Accuracy Comparison and Explanation

In this assignment I trained a neural network with two hidden layers using the MNIST handwritten digits dataset. I tested different values for the number of epochs to see how additional training time would affect the accuracy of the model. The baseline model used 20 epochs, while the experiments tested both a smaller value (5 epochs) and a larger value (40 epochs). Changing the number of epochs allows the model to pass through the training data more or fewer times, which can impact how well the network learns the patterns in the data.

When the model was trained with only 5 epochs, the training accuracy reached about 90.3%, the validation accuracy was about 91.3%, and the test accuracy was approximately 91.17%. This lower accuracy suggests the model did not have enough time to fully learn the patterns in the handwritten digits. With the baseline configuration of 20 epochs, the model performed better. The training accuracy increased to about 94.39%, the validation accuracy reached about 94.77%, and the test accuracy improved to around 94.40%. This shows that giving the model more training cycles allowed it to adjust its weights more effectively and improve its predictions.

In the final experiment, I increased the number of epochs to 40. With this additional training time, the model achieved a training accuracy of about 96.37%, validation accuracy of about 96.21%, and a test accuracy of approximately 96.19%. This indicates that the model continued to improve as it saw the training data more times. Overall, increasing the number of epochs improved the accuracy across the training, validation, and test datasets because the model had more opportunities to reduce prediction errors. However, if the number of epochs becomes too large, the model could eventually begin to overfit the training data, where performance on new data may stop improving or begin to decrease.